library/init/lean/elaborator.lean

```diff
@@ -930,6 +930,42 @@ let matchesOnly := match spec.only with
 | some only := n = only.id.val ∨ only.ids.any (λ id, n = id.val) in
 if matchesOnly then some (spec.id.val ++ n) else none
 
+/-- Global name lookup
+
+Accessible global name:
+A name `n` is called an *accessible global name* if the environment contains a
+declaration `n` or `_private.<module>.n` where `<module>` is the name of the
+current module.
+
+Active open/export command:
+A command `open/export ns ...` is *active* if
+* it is declared within the current scope or
+* it is an export command inside a namespace `ns'` and there is an active
+  open/export command for `ns'`.
+
+Outside the local context, a name `n` is looked up using the current elaborator
+state via the following formal steps:
+* If `n` matches a section variable, return the unique internal name of this
+  variable.
+* Otherwise, if there is a surrounding `namespace ns` command and `ns.n` is a
+  non-protected global accessible name, return this name. If there are multiple
+  matching namespaces, the innermost one wins.
+* Otherwise, `n` may refer to an overloaded name. Collect all matching names
+  from the following sources:
+  + add `m` if (`n = m` or `n = _root_.m`) and `m` is a global accessible name.
+  + add `ns.m` if it is not protected and there is an active open/export command
```

Comment:
I am concerned about this step. A user may (indirectly) import hundreds/thousands of .olean files, and these files may contain several `export` declarations. The stdlib already contains more than 30 `export` declarations. I imagine a big project may contain thousands of such declarations. How do we avoid this potential performance problem? Should we use a cache from `Name -> List Name`? Not sure it would be sufficient, suppose we have `N` different identifiers in a given scope, and `M` active `export` declarations. It would still take `O(N M)` even if we use a cache. Moreover, unless we use a clever caching mechanism, it will have to be reset whenever we use `open` , `export`, `namespace`, `variable` and `end` commands. BTW, note that Lean3 resolution procedure terminates quickly for all local variables, not just section variables. We can't do the same in Lean4 since we don't know what is a binder at preresolve time.

Reply:
If we want to beat `O(N M)`, I don't see any way around using the Lean 3 approach of pre-populating a set of accessible (global) names in the current context, i.e. using a forward-style implementation for this backwards-style formal search:
* On `namespace`, `open`, or `export`, the behavior should not be any different from Lean 3: we add all names currently activated by this command to the set
* On any (user-visible, non-protected) declaration (e.g. `def a.b` after `open a`), we have to check whether it is activated by any of the commands above in order to match the declarative search approach instead of Lean 3's imperative one. Now, the interesting question is where to add that code...
* On `end`, I would suggest that we simply discard the set and recompute it

Reply:
> On namespace, open, or export, the behavior should not be any different from Lean 3: we add all names currently activated by this command to the set

One thing that is bad in Lean 3 is that  we traverse **all** declarations in the environment
https://github.com/leanprover/lean4/blob/master/src/library/aliases.cpp#L148
It feels like we need an auxiliary trie-like data-structure where we branch on namespace parts. Then, to find all declarations at `foo.bla`, we start at the root, follow the `foo` branch, and then the `bla` branch.

>  On any (user-visible, non-protected) declaration (e.g. 
 after 
), we have to check whether it is activated by any of the commands above in order to match the declarative search approach instead of Lean 3's imperative one. Now, the interesting question is where to add that code...
 
Yeah. I thought about supporting user-defined callbacks in the environment. That is a function that is invoked whenever we add a new declaration. It is not great since we have to decide when to invoke it. Adding a new declaration is not the only event that would affect the set. Marking a declaration as `protected` also affects the result.

> On 
, I would suggest that we simply discard the set and recompute it

I think this one we can support using a stack/list of scopes. The same approach we used to implement scoped extensions.

BTW, we should also consider simplifying the `open/export` commands. These commands have a few features we've never used. It seems the most common uses are:
`open bla`
`open bla (id_1 ... id_n)`
`export bla (id_1 ... id_n)`
Then we have a few `hiding`s (0 at stdlib and 7 at mathlib). Example
`open filter (hiding realizer)`
And 0 renamings.

Reply:
Another feature we never seem to use is the `as` keyword.

Reply:
Yet another relevant discussion we had (also with @joehendrix) was about bringing "true" modules (in the file-level sense, like in Haskell or Rust) to Lean where each declaration is prefixed with the name of its module. Regardless of what exactly that would look like, it would most probably further increase the need for efficient resolution caches and algorithms.

Reply:
It seems we can cover 99.99999% of the use cases with 
```
open bla
open bla (id_1 ... id_n)
export bla (id_1 ... id_n)
open bla (hiding id_1 ... id_n)
```
The only one of these commands where the Lean3 imperative approach is deficient is `open bla`. We can handle it by having a list of opened namespaces. For `open bla (id_1 ... id_n)` and `export bla (id_1 ... id_n)` we create the aliases and we are done. For `open bla (hiding id_1 ... id_n)`  we claim it is syntax sugar for `open bla (set of identifiers in bla  - {id_1, ..., id_n})`, and handle by eagerly creating aliases as we do for `open bla (id_1 ... id_n)`.

Reply:
Here is yet another alternative design. `open` is super general and we implement it as it is described at `elaborator.lean`. The number of `open` commands in a file is usually small. That is, the `M` at `O(M N)` is quite small. We restrict only `export` instead, and we only support 
```
export bla (id_1 ... id_n)
```
which is the one we use all the time. This one creates aliases eagerly, and we just have to check the alias table to handle them.

Reply:
Huh, that does sound pretty sensible to me

Reply:
IIRC, one other topic where we wanted to make use of namespacing is for collections of scoped attribute instances, be it simpsets or notation scopes. If we want to enable users to e.g. deactivate built-in syntax (in order to replace it with a custom implementation), I imagined this would be done via
```
hide Init.Lean.Parser.Command.structure
```
(`hide` should also be added to the description of name lookup)
To do that, all `[commandParser]` instances inside that namespace would have to be scoped, and we should of course do an
```
export [commandParser] Init.Lean.Parser.Command
```
after the namespace to make them available by default. However, now I wonder about alternative designs that avoid this unqualified `export`.
* Do not make the instances scoped and use
```
local attribute [-commandParser] Init.Lean.Parser.Command.structure
```
instead. Bad when you want to deactivate _all_ parsers/simp lemmas of a collection (`hide Init.Lean.Parser.Command` doesn't work either right now, but I think it should).
* Do not remove the old parser, just add the new one with a higher priority. Doesn't help with removing simp lemmas.
* Do not avoid unqualified `export`s but stipulate that they only export the identifiers _currently_ in the namespace, like you suggested for `hiding`.

It's difficult to imagine what exactly users will need from the future system. What do you think?

Reply:
> It's difficult to imagine what exactly users will need from the future system. What do you think?

In the near future, we should support the scenarios we really care. For parser extensions, there are two scenarios I want to support asap:

1- Scoped extensions that are activated using `open <name>` where `<name>` is a namespace containing the scoped attributes.

2- Definition of new non-terminals (and DSLs). Each non-terminal is a new attribute which comes equipped with its own parsing tables.

It is not hard to support persistent deletion for parser attributes. That being said, I suspect it will be a mess if let users do it. If users want to design a radically different syntax, they should define a new parsing attribute. The new attribute would correspond to a new embedded DSL. 

For `[simp]`, I just want to be able to have scoped extensions.